src/hooks/useComputeDestinationTokens.ts

```diff
@@ -13,95 +14,119 @@ type Props = {
   sourceChain: Chain | undefined;
   sourceToken: Token | undefined;
   destChain: Chain | undefined;
-  route?: string;
 };
 
 type ReturnProps = {
   supportedDestTokens: Token[];
   isFetching: boolean;
 };
 
+/**
+ * Fetch supported destination token IDs from routes.
+ * Memoized to avoid redundant API calls.
+ */
+const fetchSupportedDestTokenIds = memoize(
+  async (
+    sourceTokenKey: string | undefined,
+    sourceChain: Chain,
+    destChain: Chain,
+  ): Promise<TokenId[]> => {
+    // Find the actual token from the key
+    const sourceToken = sourceTokenKey
+      ? config.tokens.getAll().find((t) => t.key === sourceTokenKey)
+      : undefined;
+
+    try {
+      return await config.routes.allSupportedDestTokens(
+        sourceToken,
+        sourceChain,
+        destChain,
+      );
+    } catch (e) {
+      console.error(e);
+      return [];
+    }
+  },
+);
+
+/**
+ * Compute the destination tokens for a given source token and chains.
+ * This handles three cases:
+ * 1. No source chain selected - returns all tokens on destination chain
+ * 2. Both chains selected with source token - fetches supported destination tokens from routes
+ * 3. Both chains selected without source token - returns empty array
+ */
+const computeDestTokensForChains = async (
+  sourceChain: Chain | undefined,
+  destChain: Chain | undefined,
+  sourceToken: Token | undefined,
+  getOrFetchToken: (tokenId: TokenId) => Promise<Token | undefined>,
+): Promise<Token[]> => {
+  if (!destChain) {
+    return [];
+  }
+
+  // User hasn't selected a source chain yet, so we
+  // return all of the known tokens on the destination chain.
+  if (!sourceChain) {
+    return config.tokens.getAllForChain(destChain);
+  }
+
+  // Both chains selected - fetch supported tokens from routes
+  const supportedIds = await fetchSupportedDestTokenIds(
+    sourceToken?.key,
+    sourceChain,
+    destChain,
+  );
+
+  const supported: Token[] = [];
+  await Promise.all(
+    supportedIds.map(async (tokenId) => {
+      const t = await getOrFetchToken(tokenId);
```

Comment:
This is the code that seems to add tokens to the config, can there be a race condition of this adding to the config and we reading config.tokens in other calls?